File: module/gongowl.py
import openai
from config import config

openai.api_key = config.plats_api_key


class Driver:
    def __init__(self):        
        self.ChatCompletion = openai.ChatCompletion()
        self.result = {}

    def addMessages(self, userMessages) :
        
        messages=[
        {"role": "system", "content": """당신은 반드시 '안내', '입시', '진로', '기타'로만 대답합니다.
        당신의 일은 질문을 분류하는 것입니다.
        '안내'는 당신에 대한 질문입니다. 또는, '이 곳은 뭐하는 곳이야' 같은 질문입니다.
        '입시'는 입학 시험(입시), 대학교에 대한 질문입니다.
        '진로'는 공부, 장래, 학업에 대한 질문입니다.
        그 외의 질문은 '기타'입니다.
        """
        },
        {"role": "user", "content": "넌 누구야?"},
        {"role": "assistant", "content": "안내"},
        {"role": "user", "content": "뭐하는 서비스야?"},
        {"role": "assistant", "content": "안내"},
        {"role": "user", "content": "서울대학교에 대한 정보를 알려줘."},
        {"role": "assistant", "content": "입시"},
        {"role": "user", "content": "입시에 대해 알려줘."},
        {"role": "assistant", "content": "입시"},
        {"role": "user", "content": "입시 정보를 알려줘."},
        {"role": "assistant", "content": "입시"},
        {"role": "user", "content": "나는 나중에 어떤 일을 해야 할까?"},
        {"role": "assistant", "content": "진로"},
        {"role": "user", "content": "난 뭘로 먹고 살아야할까?"},
        {"role": "assistant", "content": "진로"},
        {"role": "user", "content": "나 학업에 관해서 상담해줘"},
        {"role": "assistant", "content": "진로"},
        {"role": "user", "content": "나 너무 배고파."},
        {"role": "assistant", "content": "기타"},
        ]

        messages.append(
        {"role": "user", "content": userMessages}
        )
        
        response = self.ChatCompletion.create(
            model="gpt-3.5-turbo",
            messages= messages,
            max_tokens=3,
            temperature=0,
        )

        answer = str(response.choices[0].message.content)#.replace("\n", "").strip()
        print(answer)
        if answer not in ["안내", "입시", "진로", "기타"]:
            print("answer", answer)
        
            return "죄송해요. 입시나 진로에 대해 자세하게 물어봐주세요."
        
        if answer == "안내":
            response = self.anne(userMessages)

            return response
                
        if answer == "입시":
            return "입시"

        if answer == "진로":
            response = self.jinro(userMessages)

            return response
        
        return "또 다른 질문을 해보시면 어떨까요?"


    def anne(self, userMessages):
        messages=[
        {"role": "system", "content": """당신의 이름은 공부엉이입니다.
        당신은 이름을 알려주거나, "입시, 진로에 대해 물어보세요."라고 대답하면 됩니다.
        그 외에 일상 답변은 하지 않습니다. 즉, 다른 답변은 필요없습니다.
        """
        }
        ]

        messages.append(
        {"role": "user", "content": userMessages}
        )

        response = self.ChatCompletion.create(
            model="gpt-3.5-turbo",
            messages= messages,
            temperature=0.2,
            stream=True
        )
        # answer = str(response.choices[0].message.content).strip()
        # return answer

        return response

    
    def jinro(self, userMessages):
        messages=[
        {"role": "system", "content": """당신은 진로 상담가입니다.
        당신은 질문에 대해서 성심성의껏 대답합니다.
        """
        }
        ]

        messages.append(
        {"role": "user", "content": userMessages}
        )

        response = self.ChatCompletion.create(
            model="gpt-3.5-turbo",
            messages= messages,
            temperature=1,
            stream=True
        )

        # answer = str(response.choices[0].message.content).strip()
        # return answer

        return response
    

    def saveResult(self, id, result):
        print(id, result)
        self.result[id] = result


    def stream(self, id):
        def gptstream(result):
            for chunk in result:
                chunk_message = chunk['choices'][0]['delta']  # extract the message

                yield ''.join([m.get('content', '') for m in [chunk_message]])     

        try:
            value = next(gptstream(self.result[id]))
            return value
        except:
            return "stopStreaming"

File: app/static/js/gongowl.js
const rand_0_99 = Math.floor(Math.random() * 1000);
const chatBox = document.querySelector('.chat-box');
var btnEnter = 1
let intervalId;

const sendMessage = async () => {
    document.querySelector('.chat-input button').disabled = true;
    document.querySelector('.chat-input button').textContent = "Load";
    btnEnter = 0;
    
    const chatInput = document.querySelector('.chat-input input');
    
    const chatMessage = document.createElement('div');
    chatMessage.classList.add('chat-message');
    chatMessage.innerHTML = `<p>${chatInput.value}</p>`;


    if (chatInput.value.trim().length != 0) {
        const userMessages = chatInput.value;
        chatInput.value = '';

        chatBox.appendChild(chatMessage);
        chatBox.scroll(0, chatBox.scrollHeight)
    
        const astrologerMessage = document.createElement('div');
        astrologerMessage.classList.add('bodY');
        astrologerMessage.innerHTML = `
        <div class="center">
        <div class="wave"></div>
        <div class="wave"></div>
        <div class="wave"></div>
        <div class="wave"></div>
        <div class="wave"></div>
        <div class="wave"></div>
        <div class="wave"></div>
        <div class="wave"></div>
        <div class="wave"></div>
        <div class="wave"></div>
        </div>`;
        chatBox.appendChild(astrologerMessage);
        chatBox.scroll(0, chatBox.scrollHeight);

        const datas = JSON.stringify({
                userMessages: userMessages,
                id : rand_0_99

            });
        console.log(datas)

        // fetchAuthorName(datas)
        // const iter = main(datas);
        // const request = iter.next().value; /* axios.get(...)을 받는다. */
        // request.then(res => {
        //     iter.next(res); /* iter에 res를 넘겨주면서 iter을 재실행시킨다. */
        // });


        // openai_test(userMessages);


        const response = await fetch(`/gongOwlclassification/${rand_0_99}`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                userMessages: userMessages,
                id : rand_0_99
            })
        });

        const data = await response.json();
        
        const ele = document.querySelector(".bodY");
        ele.classList.remove("bodY");

        if (data == '입시') {
            
            astrologerMessage.classList.add('select-btn');
            astrologerMessage.innerHTML = 
            `<button class ='univ' onclick="univ_onclick();this.onclick = null;"> 대학교 정보 </button>
            <button class ='academy'> 학원 정보 </button>
            `;
            disableBtn()
        }
        else {
            astrologerMessage.classList.add('chat-message');
            if (data == "streaming") {
                document.querySelector('.stop-btn button').style.visibility = 'visible';

                fetchAuthorName(astrologerMessage)
            }
            else {
                astrologerMessage.innerHTML = `<p class='assistant'>${data}</p>`;
                disableBtn()
            }
        }

        chatBox.appendChild(astrologerMessage);
        chatBox.scroll(0, chatBox.scrollHeight);

    }
    else {
        disableBtn()
    }
};

document.querySelector('.chat-input button').addEventListener('click', sendMessage);
document.querySelector('.chat-input input').addEventListener('keypress', function (e) {
if (e.key === 'Enter' && btnEnter == 1) {
// code for enter
sendMessage()
}
});

function sleep(ms) {
const wakeUpTime = Date.now() + ms;
while (Date.now() < wakeUpTime) {}
}

async function openai_test(userMessages) {

var url = "https://api.openai.com/v1/chat/completions";

var xhr = new XMLHttpRequest();
xhr.open("POST", url);

xhr.setRequestHeader("Content-Type", "application/json");
xhr.setRequestHeader("Authorization", "Bearer api-key");
xhr.onreadystatechange = function () {
if (xhr.readyState === 4) {
    console.log(xhr.status);
    console.log(xhr.responseText);
    open_ai_response = xhr.responseText;
    console.log(open_ai_response);
}};

messages=[
{"role": "system", "content": "당신은 진로 상담가입니다. 당신은 질문에 대해서 성심성의껏 대답합니다."},
{"role": "user", "content": userMessages}
]

var data = `{
"model": "gpt-3.5-turbo",
"messages": ${messages},
"temperature": 1
}`;

xhr.send(data);
}

function disableBtn() {
document.querySelector('.chat-input button').disabled = false;
document.querySelector('.chat-input button').textContent = "Send";
btnEnter = 1;
}

function fetchAuthorName(astrologerMessage) {
var streamValue = ``;

function fetchData() {
    fetch(`/stream/${rand_0_99}`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
        })
    .then((response) => response.json())
    .then((data) => {  
        if (data != "stopStreaming") {
            streamValue = streamValue.concat(data)
        }
        else {
            stopFetching()
        }

        astrologerMessage.innerHTML = `<p class='assistant'>${streamValue}</p>`;
        chatBox.appendChild(astrologerMessage);
        chatBox.scroll(0, chatBox.scrollHeight);

    })
    .catch((error) => {
    console.log(error)
    });
};

function startFetching() {
    intervalId = setInterval(fetchData, 100); // delay 1 second between requests
};

function stopFetching() {
    document.querySelector('.stop-btn button').style.visibility = 'hidden';
    clearInterval(intervalId);
    disableBtn()
};

if (astrologerMessage) {
    startFetching();
}
else {
    stopFetching();
}

}

function univ_onclick(){
const astrologerMessage = document.createElement('div');
astrologerMessage.classList.add('select-box');
astrologerMessage.innerHTML = `
<select name="univ-list" size="5">
    <option value="dongguk"> 동국대학교 </option>
    <option value="gungguk"> 건국대학교 </option>
    <option value="hongik"> 홍악대학교 </option>
</select>
`;

chatBox.appendChild(astrologerMessage);
chatBox.scroll(0, chatBox.scrollHeight);
}